cli/src/main/java/com/devonfw/tools/ide/cli/IdeCompleter.java

```diff
@@ -0,0 +1,217 @@
+import org.jline.utils.AttributedString;
+
+import com.devonfw.tools.ide.commandlet.Commandlet;
+import com.devonfw.tools.ide.commandlet.ContextCommandlet;
+import com.devonfw.tools.ide.commandlet.HelpCommandlet;
+import com.devonfw.tools.ide.context.IdeContext;
+import com.devonfw.tools.ide.property.FlagProperty;
+import com.devonfw.tools.ide.property.LocaleProperty;
+import com.devonfw.tools.ide.property.Property;
+import com.devonfw.tools.ide.property.ToolProperty;
+import com.devonfw.tools.ide.property.VersionProperty;
+import com.devonfw.tools.ide.tool.ToolCommandlet;
+import com.devonfw.tools.ide.version.VersionIdentifier;
+
+/**
+ * Implements the {@link Completer} for jline3 autocompletion. Inspired by picocli
+ */
+public class IdeCompleter extends ArgumentCompleter implements Completer {
+
+  /** Checks for invalid pattern e.g. '- foo foobar' */
+  private static final String INVALID_PATTERN_PART1 = "([\\-](\\s)+.*$";
+
+  /** Checks for invalid pattern e.g. ' foo - ' or ' - foo' */
+  private static final String INVALID_PATTERN_PART2 = "([\\s][a-z]*[\\s][\\-](\\s)*.*$)";
+
+  /** Checks for invalid pattern e.g. 'get-version foobar - foobar' */
+  private static final String INVALID_PATTERN_PART3 = "([a-z][\\-][a-z]*[\\s][\\-](\\s)*.*$))";
+
+  /** Checks for invalid pattern e.g. 'foo-bar - foobar' */
+  private static final String INVALID_PATTERN_PART4 = "(^[a-z]*(\\s)[\\-](\\s)*.*$)";
+
+  /** Pattern which should stop autocompletion e.g. 'get-version mvn -' (prevents options after tool commands) */
+  private static final String INVALID_PATTERN = INVALID_PATTERN_PART1 + "|" + INVALID_PATTERN_PART2 + "|" + INVALID_PATTERN_PART3
+  + "|" + INVALID_PATTERN_PART4;
+
+  /** Pre-compiled pattern for better usage */
+  private static final Pattern patternCompiled = Pattern.compile(INVALID_PATTERN, Pattern.MULTILINE);
+
+  private final ContextCommandlet cmd;
+
+  private final IdeContext context;
+
+  private final Set<String> commandlets = new HashSet<>();
+
+  private final Set<String> toolCommandlets = new HashSet<>();
+
+  private final Set<Property<?>> commandletOptions = new HashSet<>();
+
+  public IdeCompleter(ContextCommandlet cmd, IdeContext context) {
+
+    super(NullCompleter.INSTANCE);
+    this.cmd = cmd;
+    this.context = context;
+
+    List<Property<?>> options = cmd.getProperties();
+
+    for (Property option : options) {
+      if (option instanceof FlagProperty) {
+        commandletOptions.add(option);
+      }
+    }
+
+    Collection<Commandlet> commandletCollection = context.getCommandletManager().getCommandlets();
+
+    for (Commandlet commandlet : commandletCollection) {
+      // TODO: add more logic to remove unused keyword
+      commandlets.add(commandlet.getName());
+      commandlets.add(commandlet.getKeyword());
+    }
+
+    for (Commandlet commandlet : commandletCollection) {
+      if (commandlet instanceof ToolCommandlet) {
+        // TODO: add more logic to remove unused keyword
```

Comment:
better create the next story instead of adding TODOs to `main` branch (what we do with the merge as I guess we do not want to wait for perfection with the merge as others already rely on this being merged).

Reply:
I've added a new issue and linked it in the TODO comment.